### ISSUE TITLE
Need guideline in subclassing, etc

### ISSUE THREAD

**Issue by rongcuid** [tmbdev/webdataset#70]
With current docs, I find it a bit difficult to navigate the code base when constructing custom datasets. For instance, it is not clear whether I need to deal with worker info when I try to create an `IterableDataset` that directly feeds a WDS iterator to `DataLoader`. 

Also, I am unsure if it is possible to control sampling weight in shards (for purpose of data augmentation, dealing with extremely biased data set, etc). I can probably do it with a custom `IterableDataset` class, but I will have to experiment to see whether iterators need special treatment to avoid duplicate data.

**Comment by rongcuid** [comment]
FYI, experiment shows that if I use the WDS iterators directly, I don't need to handle the worker info myself:

```ipynb
{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "145548f5-8ac1-4518-a541-58c1beb11c45",
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "import torch.utils.data as D\n",
    "\n",
    "import webdataset as wds\n",
    "\n",
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bb368a5-bfa9-4a59-9656-a59c5e37fd79",
   "metadata": {},
   "outputs": [],
   "source": [
    "writer = wds.ShardWriter(\"positive-%03d.tar\", maxcount=2)\n",
    "for i in range(10):\n",
    "    e = {\n",
    "        \"__key__\": str(i),\n",
    "        \"txt\": f\"P{i}\"\n",
    "    }\n",
    "    writer.write(e)\n",
    "\n",
    "writer.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0abc9ce7-771f-4d40-9df0-5018cf01b322",
   "metadata": {},
   "outputs": [],
   "source": [
    "writer = wds.ShardWriter(\"negative-%03d.tar\", maxcount=2)\n",
    "for i in range(10):\n",
    "    e = {\n",
    "        \"__key__\": str(i),\n",
    "        \"txt\": f\"N{i}\"\n",
    "    }\n",
    "    writer.write(e)\n",
    "\n",
    "writer.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c7db0ad-be73-4fe0-8dba-404fb3d8e847",
   "metadata": {},
   "outputs": [],
   "source": [
    "class SamplerDataset(D.IterableDataset):\n",
    "    def __init__(self):\n",
    "        self.dataset_p = wds.WebDataset(glob.glob(\"positive-*.tar\"), shardshuffle=False)\n",
    "        self.dataset_n = wds.WebDataset(glob.glob(\"negative-*.tar\"), shardshuffle=False)\n",
    "    def __iter__(self):\n",
    "        p = True\n",
    "        p_done = False\n",
    "        n_done = False\n",
    "        p_iter = iter(self.dataset_p)\n",
    "        n_iter = iter(self.dataset_n)\n",
    "        while not p_done or not n_done:\n",
    "            if p and not p_done:\n",
    "                p = False\n",
    "                try:\n",
    "                    yield next(p_iter)\n",
    "                except StopIteration:\n",
    "                    p_done = True\n",
    "            elif not p and not n_done:\n",
    "                p = True\n",
    "                try:\n",
    "                    yield next(n_iter)\n",
    "                except StopIteration:\n",
    "                    n_done = True\n",
    "            elif not p_done:\n",
    "                try:\n",
    "                    yield next(p_iter)\n",
    "                except StopIteration:\n",
    "                    p_done = True\n",
    "            elif not n_done:\n",
    "                try:\n",
    "                    yield next(n_iter)\n",
    "                except StopIteration:\n",
    "                    n_done = True\n",
    "            else:\n",
    "                raise StopIteration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1adc360d-23e4-4958-9b1f-b34b62170d0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ds = SamplerDataset()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "339de1a6-6eca-4420-93fd-57bcae8a482f",
   "metadata": {},
   "outputs": [],
   "source": [
    "dl = D.DataLoader(ds, num_workers=2)\n",
    "len(list(dl))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1277d91c-bf3f-4d46-aa72-d865492a4b8a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}

```

I really had to trace the code base to do anything, so I suggest improving documentation. When WDS works, though, it is faster than all other methods I tried, at least on my computer.

**Comment by tmbdev** [comment]
Yes, the documentation certainly needs lots of improvements. Feedback and questions are helpful for prioritizing what to document, so thanks for taking the time.

The `WebDataset` function [is really just a shorthand](https://github.com/tmbdev/webdataset/blob/1a7f72750c8dd6e9f5541769a257ac4a9f45778a/webdataset/dataset.py#L275)  for something that constructs a pipeline that yields something that's similar to a regular PyTorch dataset. 

Although there are a bunch of options, all you really need for a WebDataset is:

```
dataset = ShardList(urls, shuffle=True).then(tar_file_expander).then(group_by_keys)
```

Afterwards, the usual `dataset.shuffle(...).decode(...).to_tuple(...)` just works.

If you want to sample your shards differently, just use:

```
class MyShardList(IterableDataset, Composable):
    def __iter__(self): ...
    def __len__(self): ...

dataset = MyShardList(urls).then(tar_file_expander).then(group_by_keys)
```

There will probably be a way of describing datasets and specifying resampling at the shard leve using a JSON file soon, since it comes up frequently.
